crates/backend/src/rest.rs

```diff
@@ -111,24 +43,27 @@ pub struct RestBackend {
     url: Url,
     /// The client to use.
     client: Client,
-    /// The backoff implementation to use.
-    backoff: LimitRetryBackoff,
-}
-
-/// Notify function for backoff in case of error
-///
-/// # Arguments
-///
-/// * `err` - The error that occurred
-/// * `duration` - The duration of the backoff
-// We need to pass the error by value to satisfy the signature of the notify function
-// for handling errors in backoff
-#[allow(clippy::needless_pass_by_value)]
-fn notify(err: reqwest::Error, duration: Duration) {
-    warn!("Error {err} at {duration:?}, retrying");
+    /// The ``BackoffBuilder`` we use
+    backoff: ExponentialBuilder,
 }
 
 impl RestBackend {
+    // call the given operation retrying non-permanent errors and giving warnings for failed operations
+    fn retry_notify<F, T>(&self, op: F) -> Result<T, reqwest::Error>
+    where
+        F: FnMut() -> Result<T, reqwest::Error>,
+    {
+        op.retry(self.backoff)
+            .notify(|err, duration| warn!("Error {err} at {duration:?}, retrying"))
+            .when(|err| {
```

Comment:
maybe changing the order to `retry(..).when(..).notify(..)` makes it even more clear what is going on here..

Reply:
I agree @aawsome 
I added a commit to do just that.

I also added some doc to the `retry_notify` method.

On a side note, the simplification is indeed nice.
You basically re-did everything for the best. Next time if you don't have time / don't want to bother, I will be pleased to apply analog simplification recommendations.